### PR TITLE
Fixes #35347 - purge old foreman_docker references

### PR DESCRIPTION
**Pull request by ianballou**
In some cases, it seems that foreman-docker records are being left behind. Most notably, `::Container` and `::DockerRegistry` permissions. There may be others, but it's difficult to reproduce since the plugin is so old.  The fix that I included in a new rake task has been tested in a user environment that exhibited the bug symptoms.

I've also added the cleanup task to the list of upgrade tasks.

There is an old cleanup script, but I'm not sure how useful it is now since it's from 2019... https://github.com/theforeman/foreman_docker/blob/master/lib/tasks/cleanup.rake

### CODE REVIEW

**Comment by ianballou** [comment]
I was also requested to include `purge:trends`.

**Comment by sayan3296** [comment]
JFYI, i had requested the inclusion of trends cleanup to fix https://projects.theforeman.org/issues/32116 as well 

The only question i have is, Should the cleanups be happening as a part of db:seed or they are needed to be done as a part of db:migrate instead ?

File: db/seeds.d/112-upgrade_tasks.rb
Comment (ezr-ondrej) [inline db/seeds.d/112-upgrade_tasks.rb]: rubocop wants coma here
```suggestion
    {:name => 'purge:trends'},
```

**Comment by ezr-ondrej** [comment]
> The only question i have is, Should the cleanups be happening as a part of db:seed or they are needed to be done as a part of db:migrate instead ?

I think seeds are fine as those are run on every upgrade, right?

**Comment by sayan3296** [comment]
> > The only question i have is, Should the cleanups be happening as a part of db:seed or they are needed to be done as a part of db:migrate instead ?
> 
> I think seeds are fine as those are run on every upgrade, right?

True indeed. As long as they get executed as a part of the upgrade and clean up the stale records, We should be good. 
I raised the point earlier as I usually notice that the cleanups are being done as a part of db:migrate only.

**Comment by sjha4** [comment]
> I think seeds are fine as those are run on every upgrade, right?

I learned this the hard way but if at-least one seed file changes between 2 upgrades in foreman and all installed plugins, seed will run. If there are no updates to any seed files, upgrade does not run db:seed. So in this case, since seed file is changing, it will run at least once when this lands and then for every upgrade that any seed file is changed/added.

**Review by ekohl — COMMENTED** [review]
I'm wondering about the flow. Can't the permissions be cleaned up in a simple DB migration? UpgradeTasks run **after** the whole upgrade has ran and Foreman is running again, which means the startup will still log the warnings. You want it to run before the app starts up.

Now the only thing is that the plugin should really be absent. I think by now we can assume that, but if needed we can make Foreman's packaging obsolete foreman_docker which means it's removed.

**Comment by ekohl** [comment]
> The only question i have is, Should the cleanups be happening as a part of db:seed or they are needed to be done as a part of db:migrate instead ?

This placement is slightly misleading. It does NOT run as part of db:seed, but as part of `upgrade:run`. It actually requires all seeds to have ran. The seeding bit is that it creates the right definitions in the DB.

**Comment by ianballou** [comment]
> Can't the permissions be cleaned up in a simple DB migration?

Yeah, they could, I built it off of the `purge` task since it was something we already had. I can make a migration and remove the upgrade task. I'll keep around the new purge command though just in case.

**Comment by ianballou** [comment]
Okay, so I pulled out the purge methods into helpers and included the helper methods in both a new migration and the old rake tasks.  I also removed the upgrade task.

**Comment by ianballou** [comment]
Oh fun, we're hitting the apipie bug in the tests now.

**Comment by ianballou** [comment]
[test katello]

File: lib/tasks/cleanup_helper.rb
Comment (ekohl) [inline lib/tasks/cleanup_helper.rb]: I get you're just moving things around, but this is odd to me. The history is still there: it happened and can help explain the past. Rails doesn't care if a schema migration is in the DB but doesn't exist on disk.

File: lib/tasks/cleanup_helper.rb
Comment (ekohl) [inline lib/tasks/cleanup_helper.rb]: Same here

File: db/migrate/20220907145552_clean_trends_and_foreman_docker_data.rb
Comment (ekohl) [inline db/migrate/20220907145552_clean_trends_and_foreman_docker_data.rb]: I think in packaging we should make foreman.rpm obsolete the foreman_docker package so we're sure it's not installed anywhere.

File: db/migrate/20220907145552_clean_trends_and_foreman_docker_data.rb
Comment (ekohl) [inline db/migrate/20220907145552_clean_trends_and_foreman_docker_data.rb]: I don't like including a module. Feels like it ends up in the global namespace. Can't it be a class with only static methods?

File: lib/tasks/cleanup_helper.rb
Comment (ianballou) [reply lib/tasks/cleanup_helper.rb]: @ezr-ondrej do you recall why the `SchemaMigration` records needed to be deleted?

If it's not immediately obvious that we can delete it, I'll leave the two cleanup methods as-is.

File: db/migrate/20220907145552_clean_trends_and_foreman_docker_data.rb
Comment (ianballou) [reply db/migrate/20220907145552_clean_trends_and_foreman_docker_data.rb]: I agree to that! https://projects.theforeman.org/issues/35538

**Comment by ianballou** [comment]
@ekohl I've made some updates and retested that the migration + rake tasks still run.

File: lib/tasks/cleanup_helper.rb
Comment (ekohl) [reply lib/tasks/cleanup_helper.rb]: I'd be ok with that.

File: lib/tasks/cleanup_helper.rb
Comment (ekohl) [inline lib/tasks/cleanup_helper.rb]: I'm wondering about layout. Should this be a service perhaps? I'm not sure I like including something from `lib/tasks` in a migration.

Comment (ianballou) [reply lib/tasks/cleanup_helper.rb]: A service makes me think it's talking to something else. `app/helpers/` seems like a decent spot perhaps

Comment (ekohl) [reply lib/tasks/cleanup_helper.rb]: Don't helpers have a specific meaning within Rails? There's the method `helper :helper` which I think loads it, but I'll admit I'm a bit fuzzy on the details now.

Comment (ianballou) [reply lib/tasks/cleanup_helper.rb]: Hm yeah I think you're right, looks like this module will get loaded in. I'll look if the services dir makes sense

Comment (ianballou) [reply lib/tasks/cleanup_helper.rb]: Okay, so it feels like these methods should be outside of `app/` since they're only rake task and migration related. So, I stuck it in `lib/` where the `seeds_helper` lives. Seemed appropriate.

**Review by ekohl — APPROVED** [review]

**Comment by ianballou** [comment]
@ekohl I don't have merge permissions

**Comment by jeremylenz** [comment]
Thanks @ianballou! merging based on @ekohl's approval

File: lib/tasks/cleanup_helper.rb
Comment (ezr-ondrej) [reply lib/tasks/cleanup_helper.rb]: sorry for late reply. We have moved the migrations under the same numbers to puppet plugin, so we need to remove them in order to be able to install the plugin again in the future.
This clean-up effectivelly undo those, so it seems natural to remove them from history (as it is basically roll back of those)

File: lib/tasks/cleanup_helper.rb
Comment (ezr-ondrej) [reply lib/tasks/cleanup_helper.rb]: ^^

File: lib/tasks/cleanup_helper.rb
Comment (ekohl) [reply lib/tasks/cleanup_helper.rb]: Thanks for that information. I think that would have been a good code comment.

**Comment by ekohl** [comment]
@ezr-ondrej pointed out that this breaks the foreman_statistics plugin, since it's unconditionally removed.

**Comment by ezr-ondrej** [comment]
I've opened a fix: #9559 thanks @ekohl for opening the ticket :)

**Pull request merged**
